[libc++] P2747R2: `constexpr` placement new (library part)

frederick-vs-ja opened the pull request:
Paper link: https://wg21.link/P2747R2

The library changes affect direct `operator new` and `operator new[]` calls even when the core language changes are absent.

The changes are not available for MS ABI because the `operator new` and `operator new[]` are from VCRuntime's `<vcruntime_new.h>`.
- Feature request submitted: https://developercommunity.visualstudio.com/t/constexpr-for-placement-operator-newope/10730304.

Notes for actual placement-new in constant evaluation:
- Clang doesn't call non-allocating `operator new`/`operator new[]` in constant evaluation (#104586).
- GCC requires the non-allocating `operator new`/`operator new[]` to be `constexpr` (https://github.com/gcc-mirror/gcc/commit/afa3a4a52cf91485477e4aaa5f05987ec7ff869d).

Drive-by: Formatting the whole `new.pass.cpp` and `new_array.pass.cpp`.

Closes #105427.

frederick-vs-ja commented on libcxx/docs/Status/Cxx2cPapers.csv:
```diff
@@ -63,7 +63,7 @@
 "`P3029R1 <https://wg21.link/P3029R1>`__","Better ``mdspan``'s CTAD","2024-03 (Tokyo)","|Complete|","19.0",""
 "","","","","",""
-"`P2747R2 <https://wg21.link/P2747R2>`__","``constexpr`` placement new","2024-06 (St. Louis)","","",""
+"`P2747R2 <https://wg21.link/P2747R2>`__","``constexpr`` placement new","2024-06 (St. Louis)","|Complete|","20.0",""
```
Should we add some notes that only the library part is implemented?

MitalAshok commented on libcxx/include/version:
```diff
@@ -510,7 +510,9 @@ __cpp_lib_void_t                                        201411L <type_traits>
 # define __cpp_lib_bind_front                           202306L
 # define __cpp_lib_bitset                               202306L
-// # define __cpp_lib_constexpr_new                        202406L
+# if !defined(_LIBCPP_ABI_VCRUNTIME)
```
Do we want a check for `__cpp_constexpr >= 202406L` here too? `constexpr operator new` has been valid since C++23 but isn't usable without compiler support, and users might check `defined(__cpp_lib_constexpr_new)` instead of `defined(__cpp_lib_constexpr_new) && __cpp_constexpr >= 202406L`

frederick-vs-ja replied:
> Do we want a check for `__cpp_constexpr >= 202406L` here too? `constexpr operator new` has been valid since C++23 but isn't usable without compiler support,

I'm pretty sure that `constexpr operator new` in this paper doesn't require any constexpr feature after C++11. So the library changes don't require any chang in the compiler.

> and users might check `defined(__cpp_lib_constexpr_new)` instead of `defined(__cpp_lib_constexpr_new) && __cpp_constexpr >= 202406L`

I don't think it's meaningful to check `__cpp_lib_constexpr_new`. The approach for the core pary in #104586 doesn't require library changes.

In other words, under Clang's current (WIP) approach for constexpr placement new, the core and library changes are independent.

MitalAshok replied:
Your right, I was thinking that `operator new` could never be called in a constant expression, which might have been ill-formed NDR before C++23, but I forgot it could just be called directly `operator new(1zu, p)` without a new-expression. Which I guess is the only thing this patch adds support for with clang.

Looks like the GCC implementation *does* check if `operator new` is marked `constexpr`: https://godbolt.org/z/hYaYYvaKc so this is needed for GCC at least.

frederick-vs-ja committed:
[libc++] P2747R2: `constexpr` placement new (library part)

Paper link: https://wg21.link/P2747R2

The library changes affect direct `operator new` and `operator new[]`
calls even when the core language changes are absent. The approach of
Clang for "actual" placement new in constant evaluation doesn't depend
on the library changes.

The changes are not available for MS ABI because the `operator new` and
`operator new[]` are from VCRuntime's `<vcruntime_new.h>`.

Drive-by: formatting the whole `new.pass.cpp` and `new_array.pass.cpp`.

ldionne commented on libcxx/include/new:
```diff
@@ -242,8 +242,14 @@ _LIBCPP_OVERRIDABLE_FUNC_VIS void operator delete[](void* __p, std::size_t __sz,
 
-_LIBCPP_NODISCARD inline _LIBCPP_HIDE_FROM_ABI void* operator new(std::size_t, void* __p) _NOEXCEPT { return __p; }
-_LIBCPP_NODISCARD inline _LIBCPP_HIDE_FROM_ABI void* operator new[](std::size_t, void* __p) _NOEXCEPT { return __p; }
+_LIBCPP_NODISCARD inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX26 void*
```
Let's update the synopsis at the top of this file accordingly.

ldionne commented on libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.placement/new.pass.cpp:
```diff
@@ -15,19 +15,30 @@
-{
-    char buf[sizeof(A)];
+TEST_CONSTEXPR_OPERATOR_NEW void test_direct_call() {
+  assert(::operator new(sizeof(int), &A_constructed) == &A_constructed);
```
```suggestion
  assert(::operator new(sizeof(A), &A_constructed) == &A_constructed);
```

Isn't that less confusing? Any reason you used `sizeof(int)`?

frederick-vs-ja replied:
> Any reason you used `sizeof(int)`?

The reason is that this line (ditto in `new_array.pass.cpp`) tested taking address of the global variable (of static storage duration) `A_constructed` whose type is `int`, not any `A` object.

ldionne replied:
Ack, I read too fast. Thanks!

ldionne reviewed (commented):
Mostly LGTM but I have a few questions.

frederick-vs-ja committed:
Merge branch 'main' into p2747r2

frederick-vs-ja committed:
Update synopsis comments to mention `constexpr`

ldionne committed:
Merge branch 'main' into p2747r2

cumcum8197 commented:
354a87445e7761de58ca46cb0162950b31708335
